services/brig/src/Brig/API.hs

```diff
@@ -950,16 +950,25 @@ sitemap o = do
 
 setPropertyH :: UserId ::: ConnId ::: PropertyKey ::: JsonRequest PropertyValue -> Handler Response
 setPropertyH (u ::: c ::: k ::: req) = do
+    empty <$ (setProperty u c k =<< lazyParsePropertyValue (lazyRequestBody (fromJsonRequest req)))
+
+setProperty :: UserId -> ConnId -> PropertyKey -> PropertyValue -> Handler ()
+setProperty u c k propval = do
     maxKeyLen <- fromMaybe defMaxKeyLen <$> view (settings . propertyMaxKeyLen)
-    maxValueLen <- fromMaybe defMaxValueLen <$> view (settings . propertyMaxValueLen)
     unless (Text.compareLength (Ascii.toText (propertyKeyName k)) (fromIntegral maxKeyLen) <= EQ) $
```

Comment:
The difference here is that we already try to parse the full thing and only afterwards check that it's not too long. Without this change, the check happens first. Maybe we should leave this as it was for now and discuss it next week?

Reply:
I don't think that's the case.  There are two independent checks: one on the key (which comes from a path segment), and one on the body length.  The check on the body length is sensitive because it needs to happen before parsing, but I didn't think that flipping the order of the two independent checks mattered, so I flipped.

Yes, let's look at it again together on monday.  It's not hard to change back, or make a more conservative change.

Reply:
2a0419f7e660c5e8557f7ac1a61fc0d71437bc2b

I think now it's the same, except for more names explaining what's happening.

Before:

```haskell
setProperty :: UserId ::: ConnId ::: PropertyKey ::: JsonRequest PropertyValue -> Handler Response
setProperty (u ::: c ::: k ::: req) = do
    maxKeyLen <- fromMaybe defMaxKeyLen <$> view (settings . propertyMaxKeyLen)
    maxValueLen <- fromMaybe defMaxValueLen <$> view (settings . propertyMaxValueLen)
    unless (Text.compareLength (Ascii.toText (propertyKeyName k)) (fromIntegral maxKeyLen) <= EQ) $
        throwStd propertyKeyTooLarge
    lbs <- Lazy.take (maxValueLen + 1) <$> liftIO (lazyRequestBody (fromJsonRequest req))
    unless (Lazy.length lbs <= maxValueLen) $
        throwStd propertyValueTooLarge
    val <- hoistEither $ fmapL (StdError . badRequest . pack) (eitherDecode lbs)
    API.setProperty u c k val !>> propDataError
    return empty
```

After:

```haskell
setPropertyH :: UserId ::: ConnId ::: PropertyKey ::: JsonRequest PropertyValue -> Handler Response
setPropertyH (u ::: c ::: k ::: req) = do
    propkey <- safeParsePropertyKey k
    propval <- safeParsePropertyValue (lazyRequestBody (fromJsonRequest req))
    empty <$ setProperty u c propkey propval

setProperty :: UserId -> ConnId -> PropertyKey -> PropertyValue -> Handler ()
setProperty u c propkey propval = do
    API.setProperty u c propkey propval !>> propDataError

safeParsePropertyKey :: PropertyKey -> Handler PropertyKey
safeParsePropertyKey k = do
    maxKeyLen <- fromMaybe defMaxKeyLen <$> view (settings . propertyMaxKeyLen)
    unless (Text.compareLength (Ascii.toText (propertyKeyName k)) (fromIntegral maxKeyLen) <= EQ) $
        throwStd propertyKeyTooLarge
    pure k

-- | Parse a 'PropertyValue' from a bytestring.  This is different from 'FromJSON' in that
-- checks the byte size of the input, and fails *without consuming all of it* if that size
-- exceeds the settings.
safeParsePropertyValue :: IO Lazy.ByteString -> Handler PropertyValue
safeParsePropertyValue lreqbody = do
    maxValueLen <- fromMaybe defMaxValueLen <$> view (settings . propertyMaxValueLen)
    lbs <- Lazy.take (maxValueLen + 1) <$> liftIO lreqbody
    unless (Lazy.length lbs <= maxValueLen) $
        throwStd propertyValueTooLarge
    hoistEither $ fmapL (StdError . badRequest . pack) (eitherDecode lbs)
```

Reply:
Ah, yes, you are right. And I also like the updated version of the code better. :+1: